### PR TITLE
Dragable mixin for components

### PR DESCRIPTION
**Pull request by spydon**
# Description

Like `Tapable`, but for components to handle drags

Dragable

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# Checklist:

If something is unclear, please submit the PR anyways and ask about what you thought was unclear.

- [x] This branch is based on `v1.0.0`
- [x] This PR is targeted to merge into `v1.0.0` (not `master` or `develop`)
- [x] I have added an entry under `[next]` in `CHANGELOG.md`
- [x] I have formatted my code with `flutter format`
- [x] I have made corresponding changes to the documentation
- [x] I have added examples for new features in `doc/examples`
- [x] The continuous integration (CI) is passing

### CODE REVIEW

File: lib/game/widget_builder.dart [hunk 1]
```diff
@@ -68,19 +85,18 @@ Widget _applyAdvancedGesturesDetectors(Game game, Widget child) {
...
         GestureRecognizerFactoryWithHandlers<
             ImmediateMultiDragGestureRecognizer>(
       () => ImmediateMultiDragGestureRecognizer(),
-      (ImmediateMultiDragGestureRecognizer instance) {
+      (MultiDragGestureRecognizer instance) {
         instance
           ..onStart = (Offset o) {
             final drag = DragEvent();
+            // TODO: Remove any padding/margin on the widget here
             drag.initialPosition = o;
```
Comment (spydon) [inline lib/game/widget_builder.dart:98]: This is a problem, `initialPosition` here needs to take into consideration that there is padding outside of the flame widget, but since there is no local position here does anyone have a clue how to do that?

**Review by renancaraujo — APPROVED** [review]

File: doc/examples/gestures/lib/main_dragables.dart [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    child: game.widget,
+  );
+
+  runApp(widget);
+}
+
+class DragableSquare extends PositionComponent with Dragable {
+  @override
+  bool debugMode = true;
+  bool _isDragged = false;
```
Comment (erickzanardo) [inline doc/examples/gestures/lib/main_dragables.dart:25]: Nit: I think `_isDragging` more suitable?

File: doc/examples/gestures/lib/main_dragables.dart [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      _isDragged = false;
+    };
+    return true;
+  }
+}
+
+class MyGame extends BaseGame with HasDragableComponents {
+  MyGame() {
+    add(DragableSquare()..anchor = Anchor.topLeft);
+    //add(DragableSquare()..y = 350);
```
Comment (erickzanardo) [inline doc/examples/gestures/lib/main_dragables.dart:59]: Can this line be removed?

File: lib/game/widget_builder.dart [hunk 4]
```diff
@@ -68,19 +85,18 @@ Widget _applyAdvancedGesturesDetectors(Game game, Widget child) {
...
     gestures[ImmediateMultiDragGestureRecognizer] =
         GestureRecognizerFactoryWithHandlers<
             ImmediateMultiDragGestureRecognizer>(
       () => ImmediateMultiDragGestureRecognizer(),
-      (ImmediateMultiDragGestureRecognizer instance) {
+      (MultiDragGestureRecognizer instance) {
         instance
           ..onStart = (Offset o) {
             final drag = DragEvent();
+            // TODO: Remove any padding/margin on the widget here
```
Comment (erickzanardo) [inline lib/game/widget_builder.dart:97]: Should we though? I think we can leave this to programmer?

**Review by erickzanardo — APPROVED** [review]

File: doc/examples/gestures/lib/main_dragables.dart (hunk 2 above)
Comment (spydon) [reply doc/examples/gestures/lib/main_dragables.dart:25]: Good point

File: doc/examples/gestures/lib/main_dragables.dart (hunk 3 above)
Comment (spydon) [reply doc/examples/gestures/lib/main_dragables.dart:59]: It should be added again, forgot about it :)

File: lib/game/widget_builder.dart (hunk 4 above)
Comment (spydon) [reply lib/game/widget_builder.dart:97]: Then we would have inconsistencies in the API since they are removed for `Tapable`.

Comment (spydon) [reply lib/game/widget_builder.dart]: I don't think there is a good way to do it so I added a comment in the code and a note in the docs about the behavior.

File: lib/game/widget_builder.dart [hunk 5]
```diff
@@ -32,10 +34,15 @@ class _GenericTapEventHandler {
...
 }
 
+class _GenericDragEventHandler {
+  void Function(DragEvent details) onReceiveDrag;
+}
+
 Widget _applyAdvancedGesturesDetectors(Game game, Widget child) {
   final Map<Type, GestureRecognizerFactory> gestures = {};
 
   final List<_GenericTapEventHandler> _tapHandlers = [];
```
Comment (renancaraujo) [inline lib/game/widget_builder.dart:44]: This gonna conflict so hard with game widget lol

File: lib/widgets/sprite_widget.dart [hunk 6]
```diff
@@ -49,7 +49,7 @@ class _SpritePainter extends CustomPainter {
     final rate = min(widthRate, heightRate);
 
     final paintSize = _sprite.srcSize * rate;
-    final anchorPosition = _anchor.relativePosition;
+    final anchorPosition = _anchor.toVector2;
```
Comment (luanpotter) [inline lib/widgets/sprite_widget.dart:52]: toVector2()?
is this an unrelated change?

Comment (luanpotter) [reply lib/widgets/sprite_widget.dart:52]: ah sorry I see it now

File: lib/anchor.dart [hunk 7]
```diff
@@ -14,11 +14,11 @@ class Anchor {
   final double x;
   final double y;
 
-  Vector2 get relativePosition => Vector2(x, y);
+  Vector2 get toVector2 => Vector2(x, y);
```
Comment (luanpotter) [inline lib/anchor.dart:17]: I think this should be a method to be consistent with others: toVector2()
or could asVector2 (getter)

Comment (luanpotter) [reply lib/anchor.dart:17]: I'm ok with get relativePosition as well

File: doc/input.md [hunk 8]
```diff
@@ -103,9 +103,18 @@ class MyGame extends Game with TapDetector {
 ```
 You can also check a more complete example [here](/doc/examples/gestures).
 
-## Tapable components
+## Tapable and Dragable components
 
-Flame also offers a simple helper to make it easier to handle tap events on `PositionComponent`. By adding the `HasTapableComponents` mixin to your game, and using the mixin `Tapable` on your components can override the following methods, enabling easy to use tap events on yours components.
+A component can extend the `Tapable` and/or the `Dragable` mixins to handle taps and drags on the component.
```
Comment (luanpotter) [inline doc/input.md:108]: A BaseComponent?

File: doc/input.md [hunk 9]
```diff
@@ -103,9 +103,18 @@ class MyGame extends Game with TapDetector {
...
+
+All overridden methods return a boolean to control if the event should be passed down further along to components underneath it.
+So say that you only want your top visible component to receive a tap and not the ones underneath it, then your `onTapDown`,  `onTapUp` and `onTapCancel` implementations should return `false`, if you want the event to go through more of the components underneath then you should return `true`.
+
+The same applies if your component has children, then the event is first sent to the leaves in the children tree and then passed further down until a method returns `false`.
+
+### Tapable components
+
+Flame also offers a simple helper to make it easier to handle tap events on `PositionComponent` (and `BaseComponent`). By adding the `HasTapableComponents` mixin to your game, and using the mixin `Tapable` on your components can override the following methods, enabling easy to use tap events on your components.
```
Comment (luanpotter) [inline doc/input.md:117]: The first sentence seems to be conflicting with the previous paragraph; we just said that there are Tappable and Draggable as simple helpers to handle events on BaseComponent.

File: lib/game/widget_builder.dart (hunk 1 above)
Comment (luanpotter) [reply lib/game/widget_builder.dart:98]: I believe @renancaraujo 's PR will allow for precisely that!

File: lib/anchor.dart (hunk 7 above)
Comment (spydon) [reply lib/anchor.dart:17]: Good point, I'll go with `toVector2()` to just keep it the same all over the board.
I think `relativePosition` is quite a confusing term, "What is it in relation to, and what is the expected output?" is what I would ask myself without looking at the code.

File: doc/input.md (hunk 9 above)
Comment (spydon) [reply doc/input.md:117]: Yeah, this needs to be updated.
